File: frameworks/Dart/angel/lib/src/query/postgres.dart
import 'dart:async';
import 'package:dart_angel_benchmark/src/models/fortune.dart';
import 'package:postgres/postgres.dart';
import '../models/models.dart';
import 'querier.dart';

class PostgresQuerier implements Querier {
  final PostgreSQLConnection connection;

  PostgresQuerier(this.connection);

  static Fortune parseFortune(List row) {
    return Fortune(id: row[0], message: row[1]);
  }

  static World parseWorld(List row) {
    return World(id: row[0], randomNumber: row[1]);
  }

  @override
  Future<List<Fortune>> getFortunes() {
    return connection.query('SELECT id, message FROM fortune').then((rows) {
      return rows.map((parseFortune)).toList();
    });
  }

  @override
  Future<World> getRandomWorld() async {
    var rows = await connection
        .query('SELECT id, randomNumber FROM world ORDER BY RANDOM() LIMIT 1');
    return parseWorld(rows[0]);
  }

  @override
  Future<World> updateWorld(id, World world) async {
    await connection.query(
        'UPDATE world SET randomNumber = @randomNumber WHERE id = @id',
        substitutionValues: {'id': id, 'randomNumber': world.randomNumber});
    return world;
  }
}

File: frameworks/JavaScript/hyperexpress/database/postgres.js
import postgres from 'postgres'
import { clientOpts } from '../config.js'

const sql = postgres({ ...clientOpts, max: 1 })

export const fortunes = async () => sql`SELECT id, message FROM fortune`

export const find = async (id) => sql`SELECT id, randomNumber FROM world WHERE id = ${id}`.then((arr) => arr[0])

export const getAllWorlds = async () => sql`SELECT id, randomNumber FROM world`

export const update = async (obj) => sql`UPDATE world SET randomNumber = ${obj.randomNumber} WHERE id = ${obj.id}`
